### DISCUSSION TITLE
Load users with roles / permissions in bulk

### DISCUSSION THREAD

**Discussion by filipagh** [fief-dev/fief#160] [Ideas]
user story:
as integrator of fief i need implement logic for loading all users with their roles / permissions so admin can easily see  and change role / permissions  of users

now with admin api we can load all users but UserRead model do not contain roles / permissions data
so only way is make request for each user to load its roles, same for permissions

is possible to add UserRoles  and UserPermissions to UserRead?

**Comment by frankie567** [comment]
I purposefully avoided that in the API to limit the quantity of data to load and keep a fast and efficient.

Could you elaborate on your use-case? Why do you need to load roles/permissions of every users at once?

**Reply by filipagh** [reply]
admin need configuration page for users, there should be loaded all users (we do not expect to exceed 20 in long long future), and their roles 
then admin can pick one or many users and set/remove specific role 

something like this 
![image](https://user-images.githubusercontent.com/22474494/227869180-bcc05137-e381-4f6b-b259-addfff301037.png)



but if is ok to load each one in own request, im fine with it, but little worry about fief api ratelimiting

**Reply by frankie567** [reply]
Ok, I see!

I came up with an idea: what would you think of endpoints that would return couples of user ids and permissions/roles id? It would look like this:

```json
{
    "count": 3,
    "results": [
        {
            "user_id": "USER_1",
            "permission_id": "PERMISSION_1"
        },
        {
            "user_id": "USER_1",
            "permission_id": "PERMISSION_2"
        },
        {
            "user_id": "USER_2",
            "permission_id": "PERMISSION_1"
        }
    ]
}
```

Under the hood, it would actually be direct queries to [`UserPermission`](https://github.com/fief-dev/fief/blob/main/fief/models/user_permission.py) and [`UserRole`](https://github.com/fief-dev/fief/blob/main/fief/models/user_role.py) tables.

In your implementation you would have to call three endpoints:

* /api/users
* /api/user-permissions
* /api/user-roles

And then for each user, reconciliate its ID with its associated permissions and roles. What do you think?

**Reply by filipagh** [reply]
yes that would be great 
but also add user email 

thanks
